File: ecos-records/src/main/java/ru/citeck/ecos/records3/rest/v1/query/QueryBody.java
package ru.citeck.ecos.records3.rest.v1.query;

import ecos.com.fasterxml.jackson210.annotation.JsonInclude;
import ecos.com.fasterxml.jackson210.annotation.JsonSetter;
import ecos.com.fasterxml.jackson210.databind.JsonNode;
import ecos.com.fasterxml.jackson210.databind.node.ObjectNode;
import ecos.com.fasterxml.jackson210.databind.node.TextNode;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records3.record.op.query.dto.RecordsQuery;
import ru.citeck.ecos.records3.rest.v1.RequestBody;

import java.util.*;

@Getter
@Setter
@JsonInclude(JsonInclude.Include.NON_NULL)
@com.fasterxml.jackson.annotation.JsonInclude(com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL)
public class QueryBody extends RequestBody {

    @Nullable
    private List<RecordRef> records;
    @Nullable
    private RecordsQuery query;
    @Nullable
    private JsonNode attributes;

    private boolean rawAtts;

    public void setRecord(RecordRef record) {
        if (records == null) {
            records = new ArrayList<>();
        }
        records.add(record);
    }

    public void setAttribute(String attribute) {
        ObjectNode objNode = Json.getMapper().newObjectNode();
        objNode.set(attribute, TextNode.valueOf(attribute));
        this.attributes = objNode;
    }

    @JsonSetter
    @com.fasterxml.jackson.annotation.JsonSetter
    public void setAttributes(Object attributes) {
        JsonNode node = Json.getMapper().toJson(attributes);
        if (node.isArray()) {
            ObjectNode objNode = Json.getMapper().newObjectNode();
            node.forEach(n -> objNode.set(n.asText(), n));
            this.attributes = objNode;
        } else {
            this.attributes = node;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        QueryBody queryBody = (QueryBody) o;
        return Objects.equals(records, queryBody.records)
            && Objects.equals(query, queryBody.query)
            && Objects.equals(attributes, queryBody.attributes)
            && Objects.equals(rawAtts, queryBody.rawAtts);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
            records,
            query,
            attributes,
            rawAtts
        );
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/rest/v1/query/QueryBody.kt
package ru.citeck.ecos.records3.rest.v1.query

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonSetter
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.node.TextNode
import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import ru.citeck.ecos.records3.rest.v1.RequestBody
import ru.citeck.ecos.records3.rest.v1.RestUtils
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.util.*

@JsonInclude(JsonInclude.Include.NON_NULL)
open class QueryBody : RequestBody() {

    private var records: MutableList<EntityRef>? = null
    private var query: RecordsQuery? = null
    var attributes: DataValue = DataValue.NULL
    var rawAtts = false

    open fun setQuery(query: RecordsQuery?) {
        this.query = query
    }

    open fun getQuery(): RecordsQuery? {
        return query
    }

    fun setRecord(record: EntityRef) {
        records = (records ?: ArrayList()).apply { add(record) }
    }

    fun getRecords(): List<EntityRef>? {
        return records
    }

    fun setRecords(records: List<EntityRef>?) {
        this.records = records?.let { ArrayList(it) }
    }

    fun setAttribute(attribute: String) {
        val objNode: ObjectNode = Json.mapper.newObjectNode()
        objNode.set<JsonNode>(attribute, TextNode.valueOf(attribute))
        attributes = DataValue.create(objNode)
    }

    @JsonSetter
    fun setAttributes(attributes: Any?) {
        this.attributes = RestUtils.prepareReqAtts(attributes)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val queryBody = other as? QueryBody ?: return false
        return (
            records == queryBody.records &&
                query == queryBody.query &&
                attributes == queryBody.attributes &&
                rawAtts == queryBody.rawAtts
            )
    }

    override fun hashCode(): Int {
        return Objects.hash(
            records,
            query,
            attributes,
            rawAtts
        )
    }
}
